extmod/moddeflate.c

```diff
@@ -0,0 +1,411 @@
+    mp_obj_deflateio_t *self = data;
+    const mp_stream_p_t *stream = mp_get_stream(self->stream);
+    int err;
+    mp_uint_t ret = stream->write(self->stream, &b, 1, &err);
+    if (ret == MP_STREAM_ERROR) {
+        mp_raise_OSError(err);
+    }
+}
+
+STATIC bool deflateio_init_write(mp_obj_deflateio_t *self) {
+    if (self->write) {
+        return true;
+    }
+
+    const mp_stream_p_t *stream = mp_get_stream_raise(self->stream, MP_STREAM_OP_WRITE);
+
+    self->write = m_new_obj(mp_obj_deflateio_write_t);
+    self->write->input_len = 0;
+
+    int wbits = self->window_bits;
+    if (wbits == 0) {
+        wbits = DEFLATEIO_DEFAULT_WBITS;
+    }
+    size_t window_len = 1 << wbits;
+    self->write->window = m_new(uint8_t, window_len);
+
+    uzlib_lz77_init(&self->write->lz77, self->write->window, window_len);
+    self->write->lz77.dest_write_data = self;
+    self->write->lz77.dest_write_cb = deflateio_out_byte;
+
+    // Write header if needed.
+    mp_uint_t ret = 0;
+    int err;
+    if (self->format == DEFLATEIO_FORMAT_ZLIB) {
+        // -----CMF------  ----------FLG---------------
+        // CINFO(5) CM(3)  FLEVEL(2) FDICT(1) FCHECK(5)
+        uint8_t buf[] = { 0x08, 0x80 };     // CM=2 (deflate), FLEVEL=2 (default), FDICT=0 (no dictionary)
+        buf[0] |= MAX(self->window_bits - 8, 1) << 4;     // base-2 logarithm of the LZ77 window size, minus eight.
+        buf[1] |= 31 - ((buf[0] * 256 + buf[1]) % 31);     // (CMF*256 + FLG) % 31 == 0.
+        ret = stream->write(self->stream, buf, sizeof(buf), &err);
+
+        self->write->input_checksum = 1; // ADLER32
+    } else if (self->format == DEFLATEIO_FORMAT_GZIP) {
+        // ID1(8) ID2(8) CM(8) ---FLG--- MTIME(32) XFL(8) OS(8)
+        // FLG: x x x FCOMMENT FNAME FEXTRA FHCRC FTEXT
+        uint8_t buf[] = { 0x1f, 0x8b, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x03 };     // MTIME=0, XFL=4 (fastest), OS=3 (unix)
+        ret = stream->write(self->stream, buf, sizeof(buf), &err);
+
+        self->write->input_checksum = ~0; // CRC32
+    }
+    if (ret == MP_STREAM_ERROR) {
+        return false;
+    }
+
+    // Write starting block.
+    uzlib_start_block(&self->write->lz77);
+
+    return true;
+}
+#endif
+
+STATIC mp_obj_t deflateio_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum {
+        ARG_stream,
+        ARG_format,
+        ARG_wbits,
+        ARG_close,
+    };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_stream, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_format, MP_ARG_INT, {.u_int = DEFLATEIO_FORMAT_NONE} },
+        { MP_QSTR_wbits, MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_close, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_FALSE} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
```

Comment:
Possible make these positional-only args, to reduce code size and allow a natmod version.

Reply:
Done.